### DISCUSSION TITLE
How to "Teardown" a single OSD device that uses a journal and re-provision

### DISCUSSION THREAD

**Discussion by kyle-rf** [rook/rook#8567] [Q&A]
Hello,

Sorry if this is long winded.  I want to try and explain what I've done, but ultimately it may or may not apply to my overarching question.

**Background**
I have a cluster with 90 OSDs on three storage nodes.  About a month or so ago, one of the osd pods started crashing.  The disk is fairly new, but we assumed maybe it was bad.  Despite smartctl test output saying it was all good.  So we removed the osd from ceph.  The cluster reports 89 OSDs and is healthy.  I ran through the [teardown](https://rook.io/docs/rook/v1.7/ceph-teardown.html) process on the removed disk.

The osd prepare jobs do not seem to be kicking off anymore.  Its been 34 days since they ran last.  I deleted the job for the node the device is on and manually recreated.  It reports the disk is available, but is not provisioning it properly.  Sometime later it did end up creating the osd pod again, but it is failing to activate now because it never provisioned/initialized the disk.

I was able to restore the PV and VG metadata for the device.  But its in a bad state where it isn't associated with /dev/dm-X because those associations were also cleared from performing the teardown.  (The activate command, I believe, is looking for the /dev/mapper/ceph-UUID  or /dev/ceph-UUID to activate.  Which doesn't exist now.) We are using journaling devices so from the code I believe it originally executed the ceph-volume command with the lvm batch arguments.  That was the only branch of code that appears to involve the metadata/journaling devices.  I was trying to see if I could bring the disk back to a state where the activate command could move forward, but this is probably not worth any more time trying to recreate what is missing.

Ultimately, I want to clean up this disk and have the operator provision an osd on the device again.  The teardown process has worked well for completely cleaning up an entire storage node for reprovisioning.  However, I am not succeeding with a single device.

**Question**
How do I get this disk, that also should use journaling, reprovisioned?  I had considered that its associated osd-db device also needs to be cleaned up.  But I am hesitant that I don't entirely understand what all needs to take place.  I don't want to dump the whole db/journal device since 9 other OSDs depend on it.

**Comment by travisn** [comment]
Creating the OSDs with a metadata device really needs to be done all at once for the node so ceph-volume will know how to carve up the metadata device for all the OSDs. I'm not sure there is a way to create a single OSD with the metadata device while other OSDs are already using that metadata.

**Reply by kyle-rf** [reply]
Hi Travis!

Thanks for your input.

I have ID information for the metadata device (**ceph.db_device=/dev/ceph-960917d7-dae4-432d-9c0d-e5e3d5782218/osd-db-839ef74f-5e72-43ee-9b37-0ea21dbd755d**).  This was in the logs associated with the OSD that was removed.  I am thinking about removing that single LV from the WAL device.  So if the osd device is completely clean and available, which the osd prepare job does see the device as available.  I am thinking the system will find that this metadata device has space available to carve up a new LV and hopefully that would be enough for the system to re-provision this OSD.

Does that make sense, do you think its worth a try?

**Reply by travisn** [reply]
As a new OSD that disk would be fine by itself, but I just don't believe the ceph-volume tool will handle creating the new OSD with the metadata. You could always try, the worst that happens is that the osd prepare job fails and you can remove the device from the list again.

**Reply by kyle-rf** [reply]
I just tried this and the osd is up now, but the mappings are different and the data device is missing in vgs, lvs, lvdisplay commands too. I'm not sure if this will lead to problems later.  

It looks like a direct mapping, the **253, 60** and **253, 9** look familiar because 9 was dm-9 that the previous LV for the journal device was mapped to and 60 would was the missing dm-60 that would have existed when all devices including this one were originally provisioned.  The block mappings are the reprovisioned osd and its metadata device.  The original setup is all the symlinks to ../dm-X:

![image](https://user-images.githubusercontent.com/26102836/130268904-198edcaa-646b-4a37-b744-866656fcd0b9.png)
